Fix incremental GC thread stack/register pushing on windows

joncham opened the pull request:
Applies multiple patches from upstream to refactor how thread suspension and stack scanning is performed. The previous PR I made was incompatible with incremental mode. This was discussed on the following commit in the past but I missed the end of the discussion: https://github.com/Unity-Technologies/bdwgc/commit/3f39ea8e2e79c4d85db445aef97742eec4d56849

I applied the patches in entirety rather than trying to hand apply the fixes on our previous approach, or only the pieces we think we need.

Fixes case 1411601.

Hamayama committed:
Fix GetThreadContext stale register values use if WoW64 (Win32)

Issue #262 (bdwgc).

* misc.c [MSWIN32 && !MSWINRT_FLAVOR && !MSWIN_XBOX1]
(GC_win32_MessageBoxA): Do not define unless SMALL_CONFIG.
* misc.c [MSWIN32 && !_WIN64 && GC_WIN32_THREADS && CHECK_NOT_WOW64]
(GC_init): Do not call IsWow64Process() and GC_win32_MessageBoxA().
* win32_threads.c [I386] (isWow64): New static variable.
* win32_threads.c [I386] (GC_push_stack_for): If isWow64 then set also
CONTEXT_EXCEPTION_REQUEST and CONTEXT_SEGMENTS bits in ContextFlags;
if isWow64, and CONTEXT_EXCEPTION_REPORTING and
CONTEXT_EXCEPTION_ACTIVE are set on return from GetThreadContext then
call GetThreadSelectorEntry and use StackLimit of FS selector to set
sp local variable (instead of context.Esp); add comment.
* win32_threads.c [I386 && DEBUG_THREADS] (GC_push_stack_for): Call
GC_log_printf() to report TIB stack limit/base and the case when
CONTEXT_EXCEPTION_REQUEST is not supported.
* win32_threads.c [I386] (GC_thr_init): Set isWow64 by IsWow64Process()
if the later is available.

ivmai committed:
Refactoring of WoW64 workaround (Win32)
(code refactoring of commit 9483d5bba)

Issue #262 (bdwgc).

* include/private/gcconfig.h [I386 && (CYGWIN32 || MSWIN32)]
(WOW64_THREAD_CONTEXT_WORKAROUND): Define macro.
* win32_threads.c [!CONTEXT_EXCEPTION_ACTIVE] (CONTEXT_EXCEPTION_ACTIVE,
CONTEXT_EXCEPTION_REQUEST, CONTEXT_EXCEPTION_REPORTING): Move macro
definition upper to be before GC_suspend(); define only if
WOW64_THREAD_CONTEXT_WORKAROUND is defined (instead of I386).
* win32_threads.c (isWow64): Move static variable upper to be before
GC_suspend(); define only if WOW64_THREAD_CONTEXT_WORKAROUND.
* win32_threads.c (GET_THREAD_CONTEXT_FLAGS): New macro.
* win32_threads.c (GC_push_stack_for): Always set context.ContextFlags
to GET_THREAD_CONTEXT_FLAGS.
* win32_threads.c [I386] (GC_push_stack_for): Always store context.Esp
to sp (as the initial value).
* win32_threads.c (GC_push_stack_for): Use WoW64 workaround only
if WOW64_THREAD_CONTEXT_WORKAROUND (instead of I386).
* win32_threads.c (GC_thr_init): Set isWow64 only if
WOW64_THREAD_CONTEXT_WORKAROUND (instead of I386).

joncham committed:
Prevent GetThreadContext failure (Windows)
(a cherry-pick of commits 3f39ea8, ab3699da from 'unity-master')

Calls to GetThreadContext may fail.  Work around this by putting
access in suspend/resume loop to advance thread past problematic areas
where suspend fails.  Capture context in per thread structure at
suspend time rather than retreiving it during the push logic.

* include/private/gcconfig.h [(I386 || X86_64) && (CYGWIN32 || MSWIN32)]
(RETRY_GET_THREAD_CONTEXT): Define macro.
* win32_threads.c [RETRY_GET_THREAD_CONTEXT]
(MAX_SUSPEND_THREAD_RETRIES): Likewise.
* include/private/gcconfig.h [NO_RETRY_GET_THREAD_CONTEXT]
(RETRY_GET_THREAD_CONTEXT): Undefine macro (for test purposes).
* win32_threads.c [RETRY_GET_THREAD_CONTEXT] (GC_Thread_Rep): Add
saved_context field; add comment.
* win32_threads.c [RETRY_GET_THREAD_CONTEXT] (GC_suspend): Define
retry_cnt local variable; set ContextFlags and call GetThreadContext()
after invocation of SuspendThread(); call ResumeThread() and proceed to
SuspendThread() if GetThreadContext() failed (the limit of iterations
is MAX_SUSPEND_THREAD_RETRIES); call Sleep(0) after ResumeThread() or
failed SuspendThread() except for the first 2 iterations.
* win32_threads.c [RETRY_GET_THREAD_CONTEXT] (GC_push_stack_for):
Declare pcontext local variable (which refers to thread->saved_context);
define context as a macro (to *pcontext) instead of a local variable;
add comment; do not set ContextFlags and do not call GetThreadContext();
undefine context after last use of PUSHn().

ivmai committed:
Fix incorrect code generation by MS VC caused by excessive Thread_Rep size
(fix of commit 449eda034)

The gctest crashes were observed in debug builds produced by MS VC.

In addition, this change greatly reduces memory usage by GC_threads[]
and dll_thread_table[] (compared to the solution that stores the full
CONTEXT in GC_Thread_Rep).

* win32_threads.c (copy_ptr_regs): Declare static function.
* win32_threads.c (PUSHED_REGS_COUNT): Define macro.
* win32_threads.c [RETRY_GET_THREAD_CONTEXT]
(GC_Thread_Rep.saved_context): Move down to be the last field; change
type from CONTEXT to word[PUSHED_REGS_COUNT]; rename to context_regs.
* win32_threads.c [RETRY_GET_THREAD_CONTEXT]
(GC_Thread_Rep.context_sp): New field.
* win32_threads.c [RETRY_GET_THREAD_CONTEXT] (GC_suspend): Declare
context local variable and use it for GetThreadContext() call; if the
latter succeeds then call copy_ptr_regs(); update comment.
* win32_threads.c (copy_ptr_regs): Define static function (move
PUSH-related part of GC_push_stack_for code here); change PUSH1(reg) to
store reg to regs[]; define context as *pcontext; add assertion that
number of stored registers is equal to PUSHED_REGS_COUNT.
* win32_threads.c [WOW64_THREAD_CONTEXT_WORKAROUND] (copy_ptr_regs):
Store ContextFlags and SegFs as the first elements of regs[].
* win32_threads.c (GC_push_stack_for): Declare i local variable (set
to 0).
* win32_threads.c [RETRY_GET_THREAD_CONTEXT] (GC_push_stack_for):
Replace pcontext and context with word *regs; set sp from
thread->context_sp; remove undef context.
* win32_threads.c [!RETRY_GET_THREAD_CONTEXT] (GC_push_stack_for):
Declare regs[PUSHED_REGS_COUNT]; call copy_ptr_regs() to initialize
regs[] and sp; limit context scope to GetThreadContext() and
copy_ptr_regs() calls only.
* win32_threads.c (GC_push_stack_for): Call GC_push_one() for each
regs[] element (except for the first 2 ones if
WOW64_THREAD_CONTEXT_WORKAROUND).
* win32_threads.c [WOW64_THREAD_CONTEXT_WORKAROUND]
(GC_push_stack_for): Define ContextFlags, SegFs local variables
(the values are obtained from regs[]); use these variables instead of
context one; do not overwrite sp local variable value so that not to
use context.Esp directly (i.e. not to use context out of its scope).

ivmai committed:
Add assertion that thread is suspended when using context registers

* win32_threads.c (GC_push_stack_for): Add assertion that thread is
suspended (before using context registers) if thread is not self and
not blocked.

ivmai committed:
Prevent use of unsaved thread context registers in incremental GC (Windows)
(fix of commits 449eda034, 190e18c75)

During incremental collection GC_push_all_stacks() may be called when
the world is not stopped.  If the context registers and sp value are
saved during thread suspension then GC_push_all_stacks() gets these
values of the previous GC cycle or, even, gets all zeros instead the
context registers and sp on the first GC after the thread creation.
This commit fixes this by calling GetThreadContext in GC_push_stack_for
if thread is not suspended.

* win32_threads.c (first_thread): Refine the comment.
* win32_threads.c [!GC_NO_THREADS_DISCOVERY && RETRY_GET_THREAD_CONTEXT]
(GC_delete_gc_thread_no_free): Reset context_sp if GC_win32_dll_threads.
* win32_threads.c (GC_push_stack_for): Change the assertion expression
from thread->suspended to thread->suspended||!GC_world_stopped (the
assertion is off unless THREAD_LOCAL_ALLOC).
* win32_threads.c [RETRY_GET_THREAD_CONTEXT] (GC_push_stack_for): If
thread is not suspended then call GetThreadContext(), if it fails then
use sp value and the context registers saved during the latest
stop-the-world; add comments.
* win32_threads.c (GC_push_all_stacks): Add the title comment (copied
from pthread_stop_world.c).

joncham committed:
Fix issue due to missing upstream commit.

We are missing commit 5668de71107022a316ee967162bc16c10754b9ce in our
branch, and previous commit references variables added in this commit.

scott-ferguson-unity commented on win32_threads.c:
```diff
@@ -1471,8 +1488,126 @@ STATIC word GC_push_stack_for(GC_thread thread, DWORD me)
+#         endif
+          GC_ASSERT(!((word)thread->stack_base
+                      COOLER_THAN (word)tib->StackBase));
+#         ifdef UNITY_MISSING_COMMIT_5668de71
```
To make sure I'm following what we're disabling here - https://github.com/Unity-Technologies/bdwgc/commit/5668de71107022a316ee967162bc16c10754b9ce adds support for manually allocated coroutine stacks - and has nothing to do with C++ coroutines, so we have no need for this?

joncham replied:
Yes, that PR adds API to manually manage thread stacks which is useful for implementing (non-Unity, general) coroutines. A subsequent PR I need modifies code from that PR. So I am left with option to:
1. Take that coroutine PR even though we don't need it
2. Delete the coroutine code from subsequent PR that I need
3. The current approach, keep the code but preprocessor define it away. My thought is this doesn't introduce risk, but makes sure we encounter this code in future if we take additional cherry-picks or merges.

TautvydasZilys commented on win32_threads.c:
```diff
@@ -1471,8 +1488,126 @@ STATIC word GC_push_stack_for(GC_thread thread, DWORD me)
+          PNT_TIB tib;
+
+          if (!GetThreadSelectorEntry(THREAD_HANDLE(thread), SegFs, &selector))
+            ABORT("GetThreadSelectorEntry failed");
```
Does this compile on UWP?

joncham replied:
It looks like the tests ran successfully, verifying now.

TautvydasZilys replied:
Great!

TautvydasZilys replied:
I was worried it might not as MSDN says it's desktop only but they might have just not updated the docs.

joncham replied:
Interesting, based on failing tests it looks like it is available on 64-bit but not 32-bit UWP.

joncham replied:
Nvm as I see this is only a code path for x86.

joncham replied:
Adjusted to cache value on thread registration rather than using this API.

TautvydasZilys commented on win32_threads.c:
```diff
@@ -2466,6 +2603,20 @@ GC_INNER void GC_thr_init(void)
+        FARPROC pfn = GetProcAddress(hK32, "IsWow64Process");
+        if (pfn
+            && !(*(BOOL (WINAPI*)(HANDLE, BOOL*))pfn)(GetCurrentProcess(),
+                                                      &isWow64))
```
1. We should call this directly. We don't support WinXP anymore;
2. Microsoft highly recommends calling `IsWow64Process2` if it's available, as `IsWow64Process` is a bit flawed.

joncham replied:
Looks like `IsWow64Process2` is Windows 10 minimum, should we dynamically lookup that then fallback to `IsWow64Process`?

joncham replied:
Actually, I don't see the benefit of using `IsWow64Process2`. Do we need the additional machine/process information?

TautvydasZilys replied:
Yes, we actually started using it within Unity recently too:

```
auto wow64Module = GetModuleHandleW(L"api-ms-win-core-wow64-l1-1-1.dll");
if (wow64Module != nullptr)
{
    typedef BOOL(WINAPI* IsWow64Process2Func)(HANDLE hProcess, USHORT* pProcessMachine, USHORT* pNativeMachine);
    auto isWow64Process2 = reinterpret_cast<IsWow64Process2Func>(GetProcAddress(wow64Module, "IsWow64Process2"));
    if (isWow64Process2 != nullptr)
    {
        ....
```

TautvydasZilys replied:
Oh, almost forgot: if you end up using it, you should forward declare `GetModuleHandleW` as while it's allowed on UWP, it is #fidef'd in the headers to "discourage" people from using it.

`extern "C" __declspec(dllimport) HMODULE WINAPI GetModuleHandleW(LPCWSTR lpModuleName);`

TautvydasZilys replied:
(we can talk on slack about it as then I can link to our source code where we use it).

joncham replied:
Resolved in latest commit.

joncham committed:
Adjust Wow64 workaround to work on UWP/WinRT.

* Dynamically lookup IsWow64Process2 and fallback to IsWow64Process if
needed.
* Store TIB on each thread and lookup via call to NtCurrentTeb on thread
registration rather than calling GetThreadSelectorEntry which is not
available on UWP.

joncham commented:
Please take a look at latest commit which allows this to work on UWP.

TautvydasZilys reviewed (approved):
Thanks! Looks great.

ivmai commented:
> Please take a look at latest commit which allows this to work on UWP.

Why do we to define WOW64_THREAD_CONTEXT_WORKAROUND (in the last commit, c511d4a) for UWP/x86?
It should not do any harm of course but it seems to me useless, unless the WoW64 bug exists in WinRT (see [GetThreadContext returns stale register values on WOW64](https://social.msdn.microsoft.com/Forums/vstudio/ja-JP/aa176c36-6624-4776-9380-1c9cf37a314e/getthreadcontext-returns-stale-register-values-on-wow64?forum=windowscompatibility) article and https://github.com/ivmai/bdwgc/issues/262 reported by @Hamayama) I doubt.

Could you please check if bdwgc works correctly on UWP/x86 even if all `define WOW64_THREAD_CONTEXT_WORKAROUND` are removed from gcconfig.h ?

ivmai commented:
>> Please take a look at latest commit which allows this to work on UWP.

> ...
> Could you please check if bdwgc works correctly on UWP/x86 even if all define WOW64_THREAD_CONTEXT_WORKAROUND are removed from gcconfig.h ?

If yes, in my understanding the whole fix should be like this:
```
diff --git a/include/private/gcconfig.h b/include/private/gcconfig.h
index 9636ea64..9a729520 100644
--- a/include/private/gcconfig.h
+++ b/include/private/gcconfig.h
@@ -1493,7 +1493,9 @@ EXTERN_C_BEGIN
 #       define GETPAGESIZE() os2_getpagesize()
 #   endif
 #   ifdef MSWIN32
-#       define WOW64_THREAD_CONTEXT_WORKAROUND
+#       ifndef MSWINRT_FLAVOR
+#         define WOW64_THREAD_CONTEXT_WORKAROUND
+#       endif
 #       define RETRY_GET_THREAD_CONTEXT
 #       define MPROTECT_VDB
 #   endif
```

TautvydasZilys commented:
> unless the WoW64 bug exists in WinRT 

We have no reason to believe this bug doesn't exist on WinRT flavor. After all, UWP applications run on the same machines as win32 applications, using the same system libraries and the kernel. The only difference is that they're sandboxed.

ivmai commented:
Okay, upstreamed (with some changes) as [b62161af](https://github.com/ivmai/bdwgc/commit/b62161afd1e4db031b81aadc869f4a76d4e77fc7)
For non-UWP, the logic remained as before (by using GetThreadSelectorEntry).
